odoo/osv/expression.py

```diff
@@ -339,6 +339,222 @@ def distribute_not(domain):
 
     return result
 
+
+def anyfy(domain, model):
+    """ Return the domain where all conditions on field sequences have been
+    transformed into 'any' conditions.
+    """
+    result = []
+    for item in domain:
+        if not is_operator(item) and not is_boolean(item):
+            left, operator, right = item
+            path = left.split('.', 1)
+            field = model._fields[path[0]]
+            if len(path) > 1 and field.relational:
+                subdomain = [(path[1], operator, right)]
+                comodel = model.env[field.comodel_name]
+                item = (path[0], 'any', anyfy(subdomain, comodel))
+            elif operator in ('any', 'not any'):
+                comodel = model.env[field.comodel_name]
+                item = (left, operator, anyfy(right, comodel))
+        result.append(item)
+    return result
+
+
+def make_tree(domain):
+    """ Return the domain as a tree, with the following structure::
+
+        <tree> ::= ('?', <boolean>)
+                |  ('!', <tree>) | ('&', [<tree]) | ('|', [<tree])
+                |  (<fieldname>, <operator>, <value)
+
+    By construction, AND (``&``) and OR (``|``) nodes are n-ary and have at
+    least two children.  Moreover, AND nodes (resp. OR nodes) do not have AND
+    nodes (resp. OR nodes) in their direct children.
+    """
+    stack = []
+    for item in reversed(domain):
+        if item == '!':
+            stack.append(tree_not(stack.pop()))
+        elif item == '&':
+            stack.append(tree_and(stack.pop(), stack.pop()))
+        elif item == '|':
+            stack.append(tree_or(stack.pop(), stack.pop()))
+        elif item == TRUE_LEAF:
+            stack.append(tree_const(True))
+        elif item == FALSE_LEAF:
+            stack.append(tree_const(False))
+        else:
+            stack.append(tuple(item))
+    return tree_and(*reversed(stack))
+
+
+def tree_const(value):
+    """ Return a boolean tree node. """
+    return ('?', value)
+
+
+def tree_not(tree):
+    """ Negate a tree node. """
+    first, second, *__ = tree
```

Comment:
isn't first an operator and second operands ?

Reply:
Not always.  `tree` is technically one of:
* `('?', bool)` (boolean constant)
* `('!', tree)` (NOT node)
* `('&', [tree, ...])` (AND node)
* `('!', [tree, ...])` (OR node)
* `(left, operator, right)` (field condition)